docs/manual/layout/templates/twig/encoding.en.md

```diff
@@ -0,0 +1,55 @@
+---
+title: "Encoding / Escaping"
+description: "Encoding and Escaping information."
+aliases:
+    - /en/layout/templates/twig/encoding/
+weight: 70
+---
+
+
+For historic reasons Contao uses *input* encoding, but Twig embraces the more
+sane *output* encoding. You can read more about the topic (and why you should
+favor output encoding) in this [OWASP article](https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#rule-0-never-insert-untrusted-data-except-in-allowed-locations) about preventing Cross Site Scripting (XSS) attacks.
+
+
+## Why you should care
+
+The gist: You, as a template designer, have to decide how things should be
+output, because *you* know the context and which content you trust or not. The
+*exact same* data can be dangerous in one context and harmless in another:
+
+With Twig we can be specific how a certain variable should be treated. Use the `|escape` or - short - `|e` 
+[filter](https://twig.symfony.com/doc/3.x/filters/escape.html) for this:
+
+```twig
+<style>
+  .box { background: {{ color|e('css') }} }
+</style>
+
+[…]
+
+<div class="box">{{ color|e('html') }}</div>
+```
+
+{{% notice note %}}
+By default Twig encodes **all** variables. The chosen escaper
+strategy will depend on the template's file extension: your `.html.twig`
+templates will automatically get the `|e('html')` treatment, so you could omit
+this part in the above example.
+{{% /notice %}}
+
+You can mark a whole section of a template to be escaped or not by using the 
+[autoescape](https://twig.symfony.com/doc/3.x/tags/autoescape.html) tag.
+
+
+## Trusted raw data
+
+If you intentionally **do** want to output a variable containing raw HTML, like 
+`<b>nice</b>`, you need to add the `|raw` [filter](https://twig.symfony.com/doc/3.x/filters/raw.html) to your variable which tells 
+Twig to skip escaping this value. Otherwise `&lt;b&gt;nice&lt;/b&gt;` will be output, i.e. a text saying *&lt;b&gt;nice&lt;/b&gt;* and 
+not a bold word <b>nice</b>. 
+
+{{% notice warning %}}
+Keep in mind, that you only ever add `|raw` to trusted input! Using `|raw` on 
+anything else may result in severe XSS vulnerabilities!
+{{% /notice %}}
```

Comment:
Thought that we should avoid insertags in templates? 
And wouldn't we have to list all the filters then? 

We want to introduce practical Twig examples in a separate PR. Maybe that would be more suitable there then

Reply:
There's a difference between using insert tags in the template (e.g. instead of a Twig filter/function or to insert something) and explicitly replacing insert tags in a piece of text.

```twig
{# Do not replace insert tags, encode #}
{{ text }}
e.g. yields "&lt;p&gt;my{{br}}text&lt;/p&gt;"

{# Do not replace insert tags, don't encode (= text raw): #}
{{ text|raw }}
e.g. yields "<p>my{{br}}text</p>"

{# Replace insert tags, encode everything: #}
{{ text|insert_tag }}
e.g. yields "&lt;p&gt;my&lt;br&gt;text&lt;/p&gt;"

{# Replace insert tags, but *only* encode the text around the insert tags (= insert tags raw): #}
{{ text|insert_tag_raw }}
e.g. yields "&lt;p&gt;my<br>text&lt;/p&gt;" (note the intact "<br>")

{# Replace insert tags, don'T encode (= all raw): #}
{{ text|insert_tag|raw }}
e.g. yields "<p><br></p>"
```

As you can see in the above example, only the template author can know if insert tags should be replaced, if they may contain HTML that will be output as such and if the text surrounding them may contain HTML that is output as such.

Reply:
Sample in last commit: https://github.com/contao/docs/pull/1082/commits/28a1bd4fb4eeb035c7d63ec631ddd3cfb9dbceb3

Reply:
Suggestion in last commit

Reply:
@m-vo But I am not sure if the examples describe the differences sufficiently and understandably?